### PR TITLE
[Packetbeat] New SIP protocol

### PR DESCRIPTION
**Pull request by marc-gr**
This is a re-opening of https://github.com/elastic/beats/pull/7181 to fix merge issues.

TODO:

- [x] Parse VIA
- [x] Parse Contact
- [x] Parse Auth
- [x] Parse SDP body

Thanks to @tj8000rpm for the original work.

From the original PR:

Hi all. I implemented a new protocol SIP to packetbeat. #152 
The SIP(Session Initiation Protocol) is a communications protocol for signaling and controlling multimedia communication sessions. SIP is used many VoIP applications at not only enterprise uses but also telecom careers.

SIP is text-base protocol like HTTP. But SIP has various unique features like :
- SIP is server-client model, but it roles may changes call by call.
- SIP is request-response model, but server may (usualy) reply many responses for one request.
- There many requests and responses in one call.
- It is not know when the call will end.

Therefore, I implemented with following plans.
- Published for each SIP message(request or response)
- Additional timestamp to make it possible to keep message order
- Parse and store each header field and SDP body(option)
- Received raw message is stored as text value(option)

More detail is writen in README.md

### CODE REVIEW

**Comment by tj8000rpm** [comment]
@marc-gr 
I apologize my late reply.🙇 
And Thank you for take over my Pull Request.I'll leave it to you. ❤️

**Comment by dainperkins** [comment]
@marc-gr  - is the json in the readme the planned final output?  (just wondering if ECS normalization is handled later, or if it should be worked in at this level?)

**Comment by marc-gr** [comment]
> @marc-gr - is the json in the readme the planned final output? (just wondering if ECS normalization is handled later, or if it should be worked in at this level?)

This is the documentation coming from the original contribution, I will update all of it once we decide on mappings at https://github.com/elastic/ecs/issues/420 and changes are done. Also will add some test pcaps to have some examples.

File: packetbeat/protos/sip/sip_message.go
Comment (andrewkroh) [inline packetbeat/protos/sip/sip_message.go:127]: Use `var cutPosS []int` instead of allocating the empty slice. https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

File: packetbeat/protos/sip/sip_message.go
Comment (andrewkroh) [inline packetbeat/protos/sip/sip_message.go:501]: After merging this implementation let's create issues for the remaining TODOs and possibly document any limitations if they are likely to be encountered.

File: packetbeat/protos/sip/sip_message.go
Comment (andrewkroh) [inline packetbeat/protos/sip/sip_message.go:505]: `s/nill/nil/g`

File: packetbeat/protos/sip/sip_plugin.go
Comment (andrewkroh) [inline packetbeat/protos/sip/sip_plugin.go:311]:
```suggestion
	defer logp.Recover("Sip ParseUDP")
```

File: packetbeat/protos/sip/_meta/fields.yml
Comment (andrewkroh) [inline packetbeat/protos/sip/_meta/fields.yml:105]: There is a `data_nanos` in ES now. I think that could be applied here. It would still need to be in the SIP namespace, but the field name could be like `sip.timestamp`. https://www.elastic.co/guide/en/elasticsearch/reference/7.9/date_nanos.html

File: packetbeat/_meta/sample_outputs/sip.json
Comment (andrewkroh) [inline packetbeat/_meta/sample_outputs/sip.json:118]: Is this generated from the latest source?

**Review by andrewkroh — COMMENTED** [review]
Thanks for helping more the SIP PR along.

- There's a PCAP file in the PR. Is there a test that uses it? If not, can we add a "system test" that uses the file.
- This will need an asciidoc file. And SIP will need added in a few places.
  - https://www.elastic.co/guide/en/beats/packetbeat/current/packetbeat-overview.html
  - https://www.elastic.co/guide/en/beats/packetbeat/current/configuration-protocols.html
- SIP should be added to packetbeat config files.

**Comment by andrewkroh** [comment]
For releasing this as beta, the code itself should log a notice about the use of beta software. We have the cfgwarn package that we normally use for this.

File: packetbeat/_meta/sample_outputs/sip.json
Comment (dainperkins) [inline packetbeat/_meta/sample_outputs/sip.json:12]: should be "authentication" = any chance the user.name could be parsed from the auth headers?

File: packetbeat/_meta/sample_outputs/sip.json
Comment (dainperkins) [inline packetbeat/_meta/sample_outputs/sip.json:62]: status should only be for sip.type "response"

File: packetbeat/_meta/sample_outputs/sip.json
Comment (dainperkins) [inline packetbeat/_meta/sample_outputs/sip.json:63]: does this need to be renamed or removed?

File: packetbeat/_meta/sample_outputs/sip.json
Comment (marc-gr) [reply packetbeat/_meta/sample_outputs/sip.json:74]: Those are used by packetbeat and checked by system tests, we also have the sip specific ones.

File: packetbeat/_meta/sample_outputs/sip.json
Comment (marc-gr) [reply packetbeat/_meta/sample_outputs/sip.json]: Same as previous one

**Comment by marc-gr** [comment]
After talking with @jamiehynds and since it is not very clear yet how we will want to deal with the multiple `Via` headers and the multiple `SDP` media types and attributes, for now I just added them as `via.original` and `sdp.body.original` to make them queryable, and to have some time to think about how we plan to deal with them once people starts using the proto more.

**Comment by webmat** [comment]
@marc-gr I haven't been able to follow all of @dainperkins' work on the SIP proposal for fields. I just skimmed it now.

The one thing that jumped out to me, that I'd like addressed in this PR is that all version fields should be keyword, not long/integer. I think even the event samples you have wouldn't work, since they have `2.0` in there. Also, we will soon transition all version fields to `version` (in the keyword family) which will enable other cool stuff.

**Comment by marc-gr** [comment]
Yeah I notices that in the doc, but could not fix it because lack of permissions, in my code they are keywords already 👍

File: packetbeat/protos/sip/parser.go
Comment (andrewstucki) [inline packetbeat/protos/sip/parser.go:447]: Is this really needed? It looks like this is trying to potentially avoid a slice allocation by modifying what might be (I haven't traced through the code fully) a reusable byte slice, `buf`, in-place. Are we worried that the volume of allocations might be so high in this parser that we'll cause extra GC/Memory overhead that impacts overall packetbeat performance? I would imagine that we wouldn't require this parser to be _so_ performance intensive where this would actually matter much...

**Review by andrewstucki — COMMENTED** [review]
Have some general questions/comments, don't know enough about the protocol itself to opine too much there, but overall looks fairly solid to me.

File: packetbeat/tests/system/golden/sip-expected.json
Comment (dainperkins) [reply packetbeat/tests/system/golden/sip-expected.json]: did you strip out the related.user?

Comment (marc-gr) [reply packetbeat/tests/system/golden/sip-expected.json]: only when the value is `-` in addition to empty

Comment (dainperkins) [reply packetbeat/tests/system/golden/sip-expected.json]: one would have thought I'd have figured that out the first time I looked at it :)

**Review by adriansr — APPROVED** [review]

**Pull request merged**
